### PR TITLE
Opengl

### PR DESCRIPTION
**Pull request by PurpurTentakel97**
implemented first stayable window.

need to move the main loop someware else but im not able to include `gl_window.hpp` and `gl_error` in `gl_initialize`.
also im not shure if its good to return the window as it is now. maybe use a `std::shared_ptr` or a `std::unique_ptr` here

CI could work when youre branche is rebased because `add(int,int)` is from coder2k`s main function.

### CODE REVIEW

**Comment by mgerhold** [comment]
I merged the other PR, so you should be able to rebase now (to make the CI work again).

**Comment by PurpurTentakel97** [comment]
CI is working again.
include issue is still a problem.
i have so idea how to make this accessibleproperly.

**Comment by PurpurTentakel97** [comment]
need some rework

File: src/silly_sprites/include/gl_initialize.hpp [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+
+namespace sly::gl {
+
+	void initialize(int width, int height);
+
+}
```
Comment (mgerhold) [inline src/silly_sprites/include/gl_initialize.hpp:7]: Missing newline at the end of the file.

File: src/silly_sprites/opengl/gl_error.hpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+namespace sly::gl {
+	
+	class WindowError final {
+    private:
+		std::string m_message;
+
+    public:
+		WindowError(std::string_view message);
+		[[nodiscard]] std::string_view get_message() const;
+	};
```
Comment (mgerhold) [inline src/silly_sprites/opengl/gl_error.hpp:14]: This is some weird indentation.

File: src/silly_sprites/opengl/gl_window.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        glfwSetFramebufferSizeCallback(window, callback);
+
+        spdlog::info("window initialized");
+
+        return window ;
+    }
+    void Window::destroy() {
+        glfwTerminate();
+        spdlog::info("window destroyed");
+    }
```
[the remark this answers is not in the document]
Comment (mgerhold) [reply src/silly_sprites/opengl/gl_window.cpp:46]: Also: Don't you ever call `glfwDestroyWindow()`? I think that's missing.

File: src/silly_sprites/opengl/gl_window.cpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
+#include "gl_window.hpp"
+#include <glad/gl.h>
+
+namespace sly::gl {
+
+    Window::Window(GLFWwindow* window) 
+        : m_window{ window } { }
+
+    tl::expected<Window, WindowError> Window::create(int const width, int const height) {
```
Comment (mgerhold) [inline src/silly_sprites/opengl/gl_window.cpp:9]:
```suggestion
    [[nodiscard]] tl::expected<Window, WindowError> Window::create(int const width, int const height) {
```

File: src/silly_sprites/opengl/gl_window.hpp [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+#include "gl_error.hpp"
+
+namespace sly::gl {
+	
+	class Window final {
+	private:
+		GLFWwindow* m_window;
+
+    public:
+		Window(GLFWwindow* window);
```
Comment (mgerhold) [inline src/silly_sprites/opengl/gl_window.hpp:16]: The constructor should be private if you want to force the user to use the `create` function.

File: src/silly_sprites/opengl/gl_window.hpp [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+		[[nodiscard]] static tl::expected<Window,WindowError> create(int width, int height);
+		static void destroy();
+
+		[[nodiscard]] GLFWwindow* get() const;
```
Comment (mgerhold) [inline src/silly_sprites/opengl/gl_window.hpp:21]: I think this function should not be "open to the public". Maybe use introduce `friend` functions or classes if they really need to access this.
Reasoning: The fact that you use a `GLFWWindow*` internally is an implementation detail. This should not be visible to the outside.

File: src/silly_sprites/opengl/gl_window.cpp [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+		auto const result{ glfwInit() };
```
Comment (mgerhold) [inline src/silly_sprites/opengl/gl_window.cpp:10]: This line is indented more than all other lines.

File: src/silly_sprites/opengl/gl_window.cpp [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        if (result == GLFW_FALSE) {
+            spdlog::critical("Failed to initialize GLFW");
+            return tl::unexpected(WindowError{ "Failed to initialize GLFW" });
+        }
+        glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
+        glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
+        glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
+
+        GLFWwindow* window = glfwCreateWindow(width, height, "coder2k bester Mann", nullptr, nullptr);
```
Comment (mgerhold) [inline src/silly_sprites/opengl/gl_window.cpp:19]: image

File: src/silly_sprites/opengl/gl_window.cpp [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        if (not window) {
```
Comment (mgerhold) [inline src/silly_sprites/opengl/gl_window.cpp:20]:
```suggestion
        if (window == nullptr) {
```

File: src/silly_sprites/opengl/gl_window.cpp [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+            spdlog::critical("Failed to create GLFW window");
+            glfwTerminate();
+            return tl::unexpected(WindowError{ "Failed to create GLFW window" });
+        }
+        glfwMakeContextCurrent(window);
+
+        if (not gladLoadGL(glfwGetProcAddress)) {
+            spdlog::critical("Failed to initialize GLSD");
+            return tl::unexpected( WindowError{ "Failed to initialize GLSD" });
```
Comment (mgerhold) [inline src/silly_sprites/opengl/gl_window.cpp:29]:
```suggestion
            spdlog::critical("Failed to initialize Glad");
            return tl::unexpected( WindowError{ "Failed to initialize Glad" });
```

File: src/silly_sprites/opengl/gl_window.cpp [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        auto const callback{[](GLFWwindow* window, int width, int height){
```
Comment (mgerhold) [inline src/silly_sprites/opengl/gl_window.cpp:32]:
```suggestion
        auto const callback{[](GLFWwindow* const window, int const width, int const height){
```

File: src/silly_sprites/opengl/gl_window.cpp [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+            glViewport(0, 0, width, height);
+        }};
+
+        callback(window, width, height);
+        glfwSetFramebufferSizeCallback(window, callback);
+
+        spdlog::info("window initialized");
+
+        return window ;
```
Comment (mgerhold) [inline src/silly_sprites/opengl/gl_window.cpp:41]:
```suggestion
        return window;
```

**Review by mgerhold — CHANGES_REQUESTED** [review]

File: src/silly_sprites/opengl/gl_window.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (PurpurTentakel97) [reply src/silly_sprites/opengl/gl_window.cpp:46]: > Also: Don't you ever call `glfwDestroyWindow()`? I think that's missing.

this was not part of the tutorial until now.
but it seems to make sense.
so I added it so the dtor.

maybe the tutorial covers that later.

File: src/silly_sprites/opengl/gl_window.hpp (hunk 5 above)
Comment (PurpurTentakel97) [reply src/silly_sprites/opengl/gl_window.hpp:16]: > The constructor should be private if you want to force the user to use the `create` function.

i'm not able to make this ctor private and delete the other ctors.
i dont see a way to return the window in `crate()` then. when an other 'ctor' is default the 'dtor' will destroy the window.

is there any workaround?

**Review by mgerhold — APPROVED** [review]

**Pull request merged**